ompi/mca/coll/xhc/coll_xhc.c

```diff
@@ -0,0 +1,700 @@
+    xhc_peer_info_t *peer_info, xhc_comm_t **comms_dst,
+    int *comm_count_dst, xhc_loc_t *hierarchy, int hierarchy_len);
+static void xhc_destroy_comms(xhc_comm_t *comms, int comm_count);
+
+static void xhc_print_info(xhc_module_t *module,
+    ompi_communicator_t *comm, xhc_data_t *data);
+
+static void *xhc_shmem_create(opal_shmem_ds_t *seg_ds, size_t size,
+    ompi_communicator_t *ompi_comm, const char *name_chr_s, int name_chr_i);
+static void *xhc_shmem_attach(opal_shmem_ds_t *seg_ds);
+static mca_smsc_endpoint_t *xhc_smsc_ep(xhc_peer_info_t *peer_info);
+
+int xhc_lazy_init(xhc_module_t *module, ompi_communicator_t *comm) {
+
+    int comm_size = ompi_comm_size(comm);
+    int rank = ompi_comm_rank(comm);
+
+    xhc_peer_info_t *peer_info = module->peer_info;
+
+    opal_shmem_ds_t *peer_cico_ds = NULL;
+    xhc_data_t *data = NULL;
+
+    int return_code = OMPI_SUCCESS;
+    int ret;
+
+    errno = 0;
+
+    // ----
+
+    /* XHC requires rank communication during its initialization.
+     * Temporarily apply the saved fallback collective modules,
+     * and restore XHC's after initialization is done. */
+    xhc_coll_fns_t xhc_fns = xhc_module_set_coll_fns(comm, module->prev_colls);
+
+    // ----
+
+    ret = xhc_module_prepare_hierarchy(module, comm);
+    if(ret != OMPI_SUCCESS) {
+        RETURN_WITH_ERROR(return_code, ret, end);
+    }
+
+    // ----
+
+    data = malloc(sizeof(xhc_data_t));
+    peer_cico_ds = malloc(comm_size * sizeof(opal_shmem_ds_t));
+    if(!data || !peer_cico_ds) {
+        RETURN_WITH_ERROR(return_code, OMPI_ERR_OUT_OF_RESOURCE, end);
+    }
+
+    *data = (xhc_data_t) {
+        .comms = NULL,
+        .comm_count = -1,
+
+        .pvt_coll_seq = 0
+    };
+
+    // ----
+
+    if(OMPI_XHC_CICO_MAX > 0) {
+        opal_shmem_ds_t cico_ds;
+
+        void *my_cico = xhc_shmem_create(&cico_ds,
+            OMPI_XHC_CICO_MAX, comm, "cico", 0);
+        if(!my_cico) {
+            RETURN_WITH_ERROR(return_code, OMPI_ERR_OUT_OF_RESOURCE, end);
+        }
+
+        /* Manually "touch" to assert allocation in local NUMA node
+         * (assuming linux's default firt-touch-alloc NUMA policy) */
+        memset(my_cico, 0, OMPI_XHC_CICO_MAX);
+
+        ret = comm->c_coll->coll_allgather(&cico_ds,
+            sizeof(opal_shmem_ds_t), MPI_BYTE, peer_cico_ds,
+            sizeof(opal_shmem_ds_t), MPI_BYTE, comm,
+            comm->c_coll->coll_allgather_module);
+        if(ret != OMPI_SUCCESS) {
+            RETURN_WITH_ERROR(return_code, ret, end);
+        }
+
+        for(int r = 0; r < comm_size; r++) {
+            peer_info[r].cico_ds = peer_cico_ds[r];
+        }
+
+        peer_info[rank].cico_buffer = my_cico;
+    }
+
+    // ----
+
+    /* An XHC communicator is created for each level of the hierarchy.
+     * The hierachy must be in an order of most-specific to most-general. */
+
+    ret = xhc_make_comms(comm, peer_info, &data->comms, &data->comm_count,
```

Comment:
It creates "XHC communicators". The "XHC communicator" term is used throughout the component -- it is the structure that groups together ranks for the internal hierarchy. Indeed there is some naming conflict with MPI communicators, though in most parts some care is taken to distinguish between the two.

I will look into renaming this function to make it more clear that it does not refer to MPI comms. Is the general existence of the "XHC communicator" term ok?

Reply:
I would avoid conflict with a preexisting term that has a well defined meaning, such as `communicator`. Why not using `context` if it really has a communication backend, or `group` if it is just a collection of processes.

Reply:
Yes I've recently started gravitating towards calling it "group". It is a group of processes, and these processes communicate amongst them in the context of the group, so it is kind of similar to an MPI communicator. But I'll take a look at renaming everything to use the term group instead which will be clearer.

Reply:
Hmmm...might want to check with @hppritcha as I believe he is already using the term "group" in this code base, as does PMIx (he uses the PMIx_Group_xxx APIs).

Reply:
If conflicting with *something* is unavoidable, may I advocate for the "XHC communicator" term:

- The hierarchy tree essentially consists of process groups: e.g. one group for all processes in the numa node, then one group to contain just the numa leaders, etc.
  - Communication takes place in the context of groups, so e.g. the group members reduce their data amongst them, then the leaders propagate it to the next group, where this repeats.
  - This is very much like an MPI communicator, a group of processes communicating with each other. In fact I imagine this structure could also be implemented using MPI communicators. So XHC communicators are like MPI comms, but hand-rolled and purpose-built.

- Wherever in XHC both MPI comms and XHC comms are in play, the MPI one is called `ompi_comm`.
- In XHC, MPI comm participants are referred to as `ranks`, whereas xhc comm participants are referred to as `members`.
- Also the typedefs generally are in the form of `xhc_comm_` not just `comm_`.

As with each framework there's some terminology that might not be immediately clear. For what it's worth I believe it's somewhat well put together :-). Of course, one might say that the framework is not coll/xhc but coll/*, and MPI comm is a core term in it.

All this is to give a complete image about the component and its use of terms, so as to make an informed decision. Either way I'm fine with calling them "XHC groups" (or something else that is a good fit) to make life easier. Which code is the one that uses `group`? The one for MPI sessions?

Reply:
> Which code is the one that uses group? The one for MPI sessions?

Yes - fwiw, I have no opinion on the matter. Just raising it to your awareness.

Reply:
Let's not mix things. A group has nothing to do with sessions, it was used well before the introduction of sessions in MPI and OMPI. It defines an ensemble of already known processes without a specialized means of communication. A communicator is an extension of groups, as it provides to one or two groups the mean of exchanging data. The notion that sessions introduced was 'process sets' (section 11.3.2), which is related to the MPI runtime and identifies the processes by their URI (Uniform Resource Identifier).

Reply:
OK, it does sounds to me like it's fitting to describe XHC's structures as a type of communicator. But if it results to too much confusion with MPI communicators, I can change the terminology. Admittedly it will require a large number of adjustments, but can of course be done if necessary. Please let me know.